Create webviews.md (#792)
tauri-apps/tauri-docs: docs/guides/webview-versions.md

[609.3.5.1.3]: https://github.com/WebKit/WebKit/blob/30fc8a44f087596c60e98adb434c0b98eccb61bb/Source/WebKit/Configurations/Version.xcconfig
[609.4.1.1.1]: https://github.com/WebKit/WebKit/blob/8df64286794c38efa4697b7c24658cb85204a070/Source/WebKit/Configurations/Version.xcconfig
[609.4.1]: https://github.com/WebKit/WebKit/blob/cb927e6151b5ef49c9ccfb13018f51471f8f1035/Source/WebKit/Configurations/Version.xcconfig
[609.2.9.1.2]: https://github.com/WebKit/WebKit/blob/ca54d252f3416c3ec64f80a084cb5c4ff7ba24f1/Source/WebKit/Configurations/Version.xcconfig
[609.1.20.111.8]: https://github.com/WebKit/WebKit/blob/5c90480a38a86464b6b421c2fd28c744b43a4faa/Source/WebKit/Configurations/Version.xcconfig
[613.3.9.1.5]: https://github.com/WebKit/WebKit/blob/7f88b99524540e94abcdef4d45c1c0324d63fb56/Source/WebKit/Configurations/Version.xcconfig

### Linux (WebKitGTK)

The fractured nature of the Linux ecosystem means it is very hard to compile accurate information about WebKitGTK on the various distros. The table below is a very incomplete list of the most commonly used distributions and their webkit versions. You should always check your distros repositories for up-to-date information.

> | Distro                                                         | webkitgtk | WebKit    | Safari Equivalent |
> | -------------------------------------------------------------------------- | --------- | --------- | ------------- |
> | Ubuntu 18.04                                                                                        | 2.20      | [606.1.4] | 12.0              |
> | Debian 9 (with backport), Debian 10                                                      | 2.24      | [608.1.6] | 13.0              |
> | Ubuntu 20.04                                                                                        | 2.28      | [610.1.1] | 14.0              |
> | Debian 11, Ubuntu 18.04 (with update), centos 8 (non-stream)             | 2.32      | [612.1.6] | 15.0              |
> | Debian 10 (with update)                                                                        | 2.34      | [613.1.1] | 15.4              |
> | Debian 11 (with update), Ubuntu 20.04 (with update), Ubuntu 22.04    | 2.36      | [614.1.6] | TP 140 (16.0)     |
> 
> [605.1.3]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.18/Source/WebKit/Configurations/Version.xcconfig
> [606.1.4]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.20/Source/WebKit/Configurations/Version.xcconfig
> [610.1.1]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.28/Source/WebKit/Configurations/Version.xcconfig
> [614.1.6]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.36/Source/WebKit/Configurations/Version.xcconfig
> [613.1.1]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.34/Source/WebKit/Configurations/Version.xcconfig
> [612.1.6]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.32/Source/WebKit/Configurations/Version.xcconfig

Reviewer: holy shit i messed up that formatting big time. github really needs monospace fonts in the suggestion dialog lol.
And i forgot the link, adding that now too:
```suggestion
| Distro                                                            | webkitgtk | WebKit    | Safari Equivalent |
| ----------------------------------------------------------------- | --------- | --------- | ----------------- |
| Ubuntu 18.04                                                      | 2.20      | [606.1.4] | 12.0              |
| Debian 9 (with backport), Debian 10                               | 2.24      | [608.1.6] | 13.0              |
| Ubuntu 20.04                                                      | 2.28      | [610.1.1] | 14.0              |
| Debian 11, Ubuntu 18.04 (with update), centos 8 (non-stream)      | 2.32      | [612.1.6] | 15.0              |
| Debian 10 (with update)                                           | 2.34      | [613.1.1] | 15.4              |
| Debian 11 (with update), Ubuntu 20.04 (with update), Ubuntu 22.04 | 2.36      | [614.1.6] | TP 140 (16.0)     |

[605.1.3]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.18/Source/WebKit/Configurations/Version.xcconfig
[606.1.4]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.20/Source/WebKit/Configurations/Version.xcconfig
[608.1.6]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.24/Source/WebKit/Configurations/Version.xcconfig
[610.1.1]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.28/Source/WebKit/Configurations/Version.xcconfig
[612.1.6]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.32/Source/WebKit/Configurations/Version.xcconfig
[613.1.1]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.34/Source/WebKit/Configurations/Version.xcconfig
[614.1.6]: https://trac.webkit.org/browser/webkit/releases/WebKitGTK/webkit-2.36/Source/WebKit/Configurations/Version.xcconfig
```

# Webview Versions

> 

Reviewer 2: Maybe have a quick intro paragraph here? Explaining how Tauri interacts with WRY to use the OS's native webview (may be able to just steal some text from somewhere in Architecture)

> ## WebView2 (Edge/Chromium)

Reviewer 2: Maybe have headings like this to make it easier to scan through? 

```suggestion
## Windows: WebView2
```

> Tauri uses WebView2 which is based on Microsoft Edge and therefore Chromium. WebView2 can update itself, you are guaranteed a relatively recent chromium build on all Windows targets.

Reviewer 2: ```suggestion
Tauri uses WebView2 which is what powers Microsoft Edge (and is built on Chromium). Since WebView2 can update itself through OS updates, you are guaranteed a relatively recent Chromium build on all Windows targets.
```

Reviewer: > what powers

Nope. they are separate, though webview2 is basicallyyy a modified edge, or _used to be_. It's a little tricky to defined now that the can hardlink each other

> through OS updates

idk, that almost sounds to much like the monthly windows updates. webview2 updates separately from that, similar to how browser can update themselves (other than safari lol)

> WebView2 is supported on Windows 7 and newer and comes preinstalled on Windows 11. On versions older than Windows 11 the installer generated by Tauri takes care of ensuring WebView2 is installed on the system.

Reviewer 2: Windows 7 32 and 64 bit right? We don't need to distinguish here?

The installer for older versions only installs webview2 if it isn't already present right?

```suggestion
WebView2 is supported on Windows 7 and newer and comes preinstalled on Windows 11. On versions older than Windows 11 the installer generated by Tauri takes care of installing WebView2 on the system if required.
```

> ## WebKit

Reviewer 2: ```suggestion
## macOS, iOS, & Linux: WebKit
```

> Tauri uses WebKit on macOS (through [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview?language=objc)) and on Linux (through webkit2gtk).

Reviewer 2: ```suggestion
Tauri uses WebKit on macOS (through [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview?language=objc)) and on Linux (through `webkit2gtk`).
```

Webkit version numbers are quite complicated, so below is some helpful information to understand them.

<details>
<summary>How to read WebKit version numbers</summary>

> WebKit version numbers are made up of 5 segments and a numeric prefix indicating the OS WebKit is built for. The prefix is called the `SYSTEM_VERSION_PREFIX` and seems to be only present for macOS and iOS builds, not for Linux. Furthermore, if the last two segments are both `0` they can be omitted; So a version like `613.2.7.0.0` would be referred to as `613.2.7`.

Reviewer 2: ```suggestion
WebKit version numbers are made up of 5 segments and a numeric prefix indicating which OS WebKit is built for. The numeric prefix is called the `SYSTEM_VERSION_PREFIX` and seems to be only present for macOS and iOS builds, not for Linux. Furthermore, if the last two segments are both `0` they can be omitted (so a version like `613.2.7.0.0` would be referred to as `613.2.7`).
```

`$(SYSTEM_VERSION_PREFIX)$(MAJOR_VERSION).$(MINOR_VERSION).$(TINY_VERSION).$(MICRO_VERSION).$(NANO_VERSION)`

### `SYSTEM_VERSION_PREFIX`

Here is what the `SYSTEM_VERSION_PREFIX` values mean:

> | macOS version | SYSTEM_VERSION_PREFIX |

Reviewer 2: Generally I try to not put any heavily formatted things (like tables) in the disclosure areas because formatting and shading makes it a bit tougher to read. Maybe just don't put this in a disclosure tag and keep it fully inlined?

| ------------- | --------------------- |
| 10.11         | 11                    |
| 10.12         | 12                    |
| 10.13         | 13                    |
| 10.14         | 14                    |
| 10.15         | 15                    |
| 11.0          | 16                    |
| 12.0          | 17                    |
| 13.0          | 18                    |

> | sdk=iphone\*  | 8                     |

Reviewer 2: Is the "*" for anything? I don't see something later in the page for it

Author: That's taken straight from the WebKit Source 🤷‍♂️ I guess we can also remove that row entirely until we have proper iOS support

### Example

The webkit version shipped with Safari 15.5 on macOS Monterey (12.x) has the following version number: `17613.2.7.1.8`.

| part                    | version |
| ----------------------- | ------- |
| `SYSTEM_VERSION_PREFIX` | 17      |
| `MAJOR_VERSION`         | 613     |
| `MINOR_VERSION`         | 2       |
| `TINY_VERSION`          | 7       |
| `MICRO_VERSION`         | 1       |

> | `NANO_VERSION`          | 8       |

Reviewer 2: Format this as a bulleted list instead of a table may be a little easier to read (since there are so many tables on the page). I'd also move this directly up under line 20 where you're listing out the sections

</details>

> The tables below map OS versions to WebKit versions and the corresponding Safari equivalent, so you can use sites like [caniuse](https://caniuse.com) to figure out if a specific web platform feature is supported.

Reviewer 2: ```suggestion
The tables below map OS versions to WebKit versions and the corresponding Safari equivalent so that you can use sites like [caniuse](https://caniuse.com) to figure out if a specific web platform feature is supported.
```

> The tables below map OS versions to WebKit versions and the corresponding Safari equivalent, so you can use sites like [caniuse](https://caniuse.com) to figure out if a specific web platform feature is supported.

Reviewer 2: Maybe move this just under the terminal command below? Between that and the table immediately after

### macOS

> On macOS Tauri uses the component that comes preinstalled with macOS since version 10.10 (Yosemite). It is considered a core component and is therefore updated with the regular OS updates. This means _unsuspported_ macOS versions also **do not** receive WebKit updates.

Reviewer 2: ```suggestion
On macOS Tauri uses the webview that comes preinstalled with macOS since version 10.10 (Yosemite). It is considered a core component and is therefore updated with the regular OS updates. This means _unsuspported_ macOS versions **do not** receive WebKit updates.
```

> To find the Webkit version used by WKWebView you can use this command:

Reviewer 2: ```suggestion
To find the Webkit version used by `WKWebView` you can use this command in the terminal on macOS:
```

```shell
awk '/CFBundleVersion/{getline;gsub(/<[^>]*>/,"");print}' /System/Library/Frameworks/WebKit.framework/Resources/Info.plist
```

> |             | macOS                            | WebKit           | Safari Equivalent | Notes                                                            |

Reviewer 2: Reverse the sort order of the table so that most recent is at the top. Can you "merge" cells vertically in a table so that all the rows in the first column are merged?

Reviewer 2: Maybe also do something similar for "Safari Equivalent" where they would span multiple rows?

Author: I'm confused, what exactly do you mean?

> |             | macOS                            | WebKit           | Safari Equivalent | Notes                                                            |

Reviewer 2: Use "OS Name | OS Version | ..." for the first two columns. Maybe move the "Safari Equivalent" column before webkit and rename it to "Safari Version" (also call it WebKit Version)?

| ----------- | -------------------------------- | ---------------- | ----------------- | ---------------------------------------------------------------- |
| High Sierra | 10.13                            | [604.1.38.1.6]   | 11.0              |                                                                  |
|             | 10.13.1                          | [604.3.5]        | 11.0.1            |                                                                  |

> |             | 10.13.2                          | [604.4.7.1.3]    | 11.0.2            | 27" iMacPro's received a special webkit version: [604.4.7.10.4]  |

Reviewer 2: ```suggestion
|             | 10.13.2                          | [604.4.7.1.3]    | 11.0.2            | 27" iMac Pro received a special WebKit version: [604.4.7.10.4]  |
```

> |             | 10.13.2 Supplemental Update      | [604.4.7.1.6]    | 11.0.2            | 27" iMacPro's received a special webkit version: [604.4.7.10.6]  |

Reviewer 2: Grammatically these suggestions are weird to not have the plural and ownership, but that's Apple's guides 🤷🏽‍♂️

```suggestion
|             | 10.13.2 Supplemental Update      | [604.4.7.1.6]    | 11.0.2            | 27" iMac Pro received a special WebKit version: [604.4.7.10.6]  |
```

|             | 10.15.2                          | [608.4.9.1.3]    | 13.0.4            |                                                                  |
|             | 10.15.3                          | [608.5.11]       | 13.0.5            |                                                                  |
|             | 10.15.4                          | [609.1.20.111.8] | 13.1              |                                                                  |
|             | 10.15.5                          | [609.2.9.1.2]    | 13.1.1            |                                                                  |
|             | 10.15.6                          | [609.3.5.1.3]    | 13.1.2            |                                                                  |
|             | 10.15.7                          | [609.4.1]        | 13.1.3            |                                                                  |
|             | 10.15.7 Security Update 2022-004 | [609.4.1.1.1]    |                   |                                                                  |

> | Big Sur     | 11.0                             | [610.2.11.1.3]   | 14.0.1            | Safari 14.0 was only ever available on iPhones                   |

Reviewer 2: Maybe be a bit more specific in the note and qualify it as "Safari 14.0.0"?

|             | 11.0.1                           | [610.2.11.51.8]  |                   |                                                                  |
|             | 11.1                             | [610.3.7.1.9]    | 14.0.2            |                                                                  |
|             | 11.2                             | [610.4.3.1.4]    | 14.0.3            |                                                                  |
|             | 11.2.1                           |                  |                   |                                                                  |
|             | 11.2.2                           |                  |                   |                                                                  |
|             | 11.2.3                           | [610.4.3.1.7]    |                   |                                                                  |

> |             | 11.3                             | [611.1.21.161.3] | 14.1              | 24" M1 iMac's received a special webkit version: [611.1.21.1.12] |

Reviewer 2: ```suggestion
|             | 11.3                             | [611.1.21.161.3] | 14.1              | 24" M1 iMac received a special WebKit version: [611.1.21.1.12] |
```

[610.4.3.1.4]: https://github.com/WebKit/WebKit/blob/b152d7889c786689406f203cc4eefea509a90302/Source/WebKit/Configurations/Version.xcconfig
[610.4.3.1.7]: https://github.com/WebKit/WebKit/blob/248c3283ebdec8bd8ae05d4d1d56390b0da28f27/Sour.3ce/WebKit/Configurations/Version.xcconfig
[609.3.5.1.3]: https://github.com/WebKit/WebKit/blob/30fc8a44f087596c60e98adb434c0b98eccb61bb/Source/WebKit/Configurations/Version.xcconfig
[609.4.1.1.1]: https://github.com/WebKit/WebKit/blob/8df64286794c38efa4697b7c24658cb85204a070/Source/WebKit/Configurations/Version.xcconfig
[609.4.1]: https://github.com/WebKit/WebKit/blob/cb927e6151b5ef49c9ccfb13018f51471f8f1035/Source/WebKit/Configurations/Version.xcconfig
[609.2.9.1.2]: https://github.com/WebKit/WebKit/blob/ca54d252f3416c3ec64f80a084cb5c4ff7ba24f1/Source/WebKit/Configurations/Version.xcconfig
[609.1.20.111.8]: https://github.com/WebKit/WebKit/blob/5c90480a38a86464b6b421c2fd28c744b43a4faa/Source/WebKit/Configurations/Version.xcconfig
[613.3.9.1.5]: https://github.com/WebKit/WebKit/blob/7f88b99524540e94abcdef4d45c1c0324d63fb56/Source/WebKit/Configurations/Version.xcconfig

### Linux (WebKitGTK)

> The fractured nature of the Linux ecosystem means it is very hard to compile accurate information about WebKitGTK on the various distros. The table below is a very incomplete list of the most commonly used distributions and their webkit versions. You should always check your distros repositories for up-to-date information.

Reviewer 2: ```suggestion
The diverse nature of the Linux ecosystem means it is very hard to compile accurate information about WebKitGTK on the various distros. The table below is a very incomplete list of the most commonly used distributions and their WebKit versions. You should always check your distro's repositories for up-to-date information.
```

Reviewer: diverse sounds too nice haha

Reviewer 2: Haha, you just want to add fuel to the fire! 🔥

Reviewer: I mean, i blamed the manjaro devs for our ssl cert issue today so yes i have a really special relationship with Linux lately :joy:

> #### Note on the Ubuntu situation

Reviewer 2: Possibly also make this a :::note ::: thing instead of a separate subheading? 

```suggestion
#### Note On Ubuntu
```

> On Ubuntu 18 through 20 it appears only `amd64` and `i386` machines received subsequent WebKit updates. All other supported arches remained on the initial releases' WebKit version. As you can see this results in really outdated Webkit versionson those machines. Since Ubuntu 22.04 however, it seems all arches receive WebKit updates.

Reviewer 2: ```suggestion
On Ubuntu 18 through 20 it appears only `amd64` and `i386` machines received subsequent WebKit updates. All other supported arches remained on the initial releases' WebKit version. This results in outdated WebKit versions on those versions. Since Ubuntu 22.04 however, it seems all arches receive WebKit updates.
```

| Distro                                                         | webkitgtk | WebKit    | Safari Equivalent |
| -------------------------------------------------------------- | --------- | --------- | ----------------- |

> | debian 9                                                       | 2.18      | [605.1.3] | 11.1              |

Reviewer 2: ```suggestion
| Debian 9                                                       | 2.18      | [605.1.3] | 11.1              |
```

> #### Note on the Ubuntu situation
> 
> On Ubuntu 18 through 20 it appears only `amd64` and `i386` machines received subsequent WebKit updates. All other supported arches remained on the initial releases' WebKit version. As you can see this results in really outdated Webkit versionson those machines. Since Ubuntu 22.04 however, it seems all arches receive WebKit updates.
> 
> | Distro                                                         | webkitgtk | WebKit    | Safari Equivalent |
> | -------------------------------------------------------------- | --------- | --------- | ----------------- |
> | debian 9                                                       | 2.18      | [605.1.3] | 11.1              |
> | Ubuntu 18.04 (arm64, armhf, ppc64el, s390x)                    | 2.20      | [606.1.4] | 12.0              |
> | Ubuntu 20.04 (arm64, armhf, ppc64el, s390x)                    | 2.28      | [610.1.1] | 14.0              |
> | Ubuntu 18.04 (amd64, i386), centos 8 (non-stream)              | 2.32      | [612.1.6] | 15.0              |
> | debian 10                                                      | 2.34      | [613.1.1] | 15.4              |
> | Ubuntu 20.04 (amd64, i386), Ubuntu 22.04, debian 11, debian 12 | 2.36      | [614.1.6] | TP 140 (16.0)     |

Reviewer: ```suggestion
| Distro                                                         | webkitgtk | WebKit    | Safari Equivalent |
| -------------------------------------------------------------------------- | --------- | --------- | ------------- |
| Ubuntu 18.04                                                                                        | 2.20      | [606.1.4] | 12.0              |
| Debian 9 (with backport), Debian 10                                                      | 2.24      | [608.1.6] | 13.0              |
| Ubuntu 20.04                                                                                        | 2.28      | [610.1.1] | 14.0              |
| Debian 11, Ubuntu 18.04 (with update), centos 8 (non-stream)             | 2.32      | [612.1.6] | 15.0              |
| Debian 10 (with update)                                                                        | 2.34      | [613.1.1] | 15.4              |
| Debian 11 (with update), Ubuntu 20.04 (with update), Ubuntu 22.04    | 2.36      | [614.1.6] | TP 140 (16.0)     |
```

Debian 9 is a little special. webkitgtk version on release and in the normal update channel are too old, only with the _opt-in_ backports repo teh can get a version >=2.22 (which we set as minimum version in wry, tho later versions may still work idk).
p.s. debian 9 release = 2.16 - debain 9 update = 2.18

@JonasKruckenberg Do you still remember the file you mentioned where you got the webkit version from? Can't seem to find it but i need the webkit version for webkitgtk@2.24

Author: webkitgtk@2.24 shipped webkit 608.1.6

Around mid-last year webkit started to switch to GitHub for version tracking, so for newer versions you have to check out the tags on GitHub. For older releases there is https://trac.webkit.org/browser specifically the releases "folder"

And then go to `Source/WebKit/Configurations/Version.xcconfig`

Reviewer: Thanks! ❤️ i updated the suggestion text.
